Package generation

freider opened the pull request:
Based on @anyman 's branch

anyman committed:
Generating full python packages

freider committed:
Work in progress

freider committed:
Fix broken tests by inserting extra newline and namespace

freider committed:
Tests for multi-module package generation

freider commented:
Work in progress but runs tests without problems. Needs testing on more substantial input, but this should be pretty closed to finished. Please squash and add proper commit messages before merging when done :)

freider commented on pyschema/source_generation.py:
```diff
@@ -22,6 +24,94 @@ def to_python_source(classes, indent=DEFAULT_INDENT):
+
+            for inlined in inlined_classes:
+                all_referenced = class_graph.find_descendents(inlined)
+                imported_classes |= set([c for c in all_referenced if c not in inlined_classes])
```
1. Spelling: descendents -> descendants
2. Just realized that we are adding un-needed imports here. Not a biggie to have unused imports, but only children/direct descendant are needed here, since all others will only be referenced in the imported files where they in turn are imported... Should be a small modification to fix this

freider commented on pyschema/source_generation.py:
```diff
@@ -22,6 +24,94 @@ def to_python_source(classes, indent=DEFAULT_INDENT):
+            self.write_namespace_file(namespace, module_code)
+
+
+def to_python_package(classes, target_folder, parent_package=None, indent=DEFAULT_INDENT):
```
Maybe add a docstring here :)
classes and target folder are pretty clear... The effect of parent_package might need some explaining though... (it's used to be able to do absolute imports everywhere, i.e. prepended on any generated import statements. It's not used for the actual folder layout, i.e. the parent_package directories will not be created in the target_folder...)

freider commented on test/source_generation_test.py:
```diff
@@ -142,3 +153,57 @@ class TestSelfReference(TestCase):
+        src = self.get_file_content("my/foo/bar.py")
+        expected_src = header_source() + "\n" + TestSourceConversion.correct
+        self.assertEquals(src, expected_src)
+        #TODO, verify __init__ files
```
Wasn't sure exactly what this TODO meant...

anyman replied:
Creating **init**.py files in all the folders so that we build a valid package strucute

anyman committed:
Multiple fixes

1. typo fix of descendants
2. quick fix in graph traversal, previously we never parsed subrecords
3. add __init__.py generation to package structure

anyman commented:
Should be replaced with https://github.com/spotify/pyschema/pull/13

anyman committed:
Modified so that only the directly references schemas are imported

anyman committed:
Adding docstring

freider committed:
Fix incorrect name

freider committed:
Adds support for storing named Enums in SchemaStore

freider committed:
Don't generate packages with reserved keywords as names. Use init files instead module files if generated module is not a leaf module

freider committed:
Test some aspects of enums in source generation

freider committed:
Avro parser now has experimental support for parsing Enum-only types and refer to previously parsed enums by name

freider committed:
Fix for luigi tests that broke due to global state
